system/Encryption/Handlers/OpenSSLHandler.php

```diff
@@ -0,0 +1,179 @@
+	 */
+	protected $cipher = 'AES-256-CTR';
+
+	// --------------------------------------------------------------------
+
+	/**
+	 * Initialize OpenSSL, remembering parameters
+	 *
+	 * @param array $params Configuration parameters
+	 *
+	 * @throws \CodeIgniter\Encryption\EncryptionException
+	 */
+	public function __construct($params = [])
+	{
+		parent::__construct($params);
+
+		$this->logger->info('OpenSSL handler initialized with cipher ' . $this->cipher . '.');
+	}
+
+	/**
+	 * Encrypt plaintext, with optional HMAC and base64 encoding
+	 *
+	 * @param  string $data   Input data
+	 * @param  array  $params Over-ridden parameters, specifically the key
+	 * @return string
+	 * @throws \CodeIgniter\Encryption\EncryptionException
+	 */
+	public function encrypt($data, $params = null)
+	{
+		// Allow key over-ride
+		if (! empty($params))
+		{
+			if (isset($params['key']))
+			{
+				$this->key = $params['key'];
+			}
+			else
+			{
+				$this->key = $params;
+			}
+		}
+		if (empty($this->key))
+		{
+			throw new EncryptionException('Encrypter needs a starter key.');
+		}
+
+		// derive a secret key
+		$secret = strcmp(phpversion(), '7.1.2') >= 0 ?
+				\hash_hkdf($this->digest, $this->key) :
+				\CodeIgniter\Encryption\Encryption::hkdf($this->key, $this->digest);
+
+		// basic encryption
+		$iv = ($iv_size = \openssl_cipher_iv_length($this->cipher)) ? \openssl_random_pseudo_bytes($iv_size) : null;
+
+		$data = \openssl_encrypt($data, $this->cipher, $this->secret, OPENSSL_RAW_DATA, $iv);
+
+		if ($data === false)
+		{
+			return false;
+		}
+
+		$result = $iv . $data;
+
+		$hmacKey = \hash_hmac($this->digest, $result, $this->secret, true);
+		$result  = $hmacKey . $result;
+
+		return $result;
+	}
+
+	// --------------------------------------------------------------------
+
+	/**
+	 * Decrypt ciphertext, with optional HMAC and base64 encoding
+	 *
+	 * @param  string $data   Encrypted data
+	 * @param  array  $params Over-ridden parameters, specifically the key
+	 * @return string
+	 * @throws \CodeIgniter\Encryption\EncryptionException
+	 */
+	public function decrypt($data, $params = null)
+	{
+		// Allow key over-ride
+		if (! empty($params))
+		{
+			if (isset($params['key']))
+			{
+				$this->key = $params['key'];
+			}
+			else
+			{
+				$this->key = $params;
+			}
+		}
+		if (empty($this->key))
+		{
+			throw new EncryptionException('Decrypter needs a starter key.');
+		}
+
+		// derive a secret key
+		$secret = strcmp(phpversion(), '7.1.2') >= 0 ?
+				\hash_hkdf($this->digest, $this->key) :
+				\CodeIgniter\Encryption\Encryption::hkdf($this->key, $this->digest);
+
+		$hmacLength = self::substr($this->digest, 3) / 8;
+		$hmacKey    = self::substr($data, 0, $hmacLength);
+		$data       = self::substr($data, $hmacLength);
+		$hmacCalc   = \hash_hmac($this->digest, $data, $this->secret, true);
+		if (! hash_equals($hmacKey, $hmacCalc))
+		{
+			throw new \CodeIgniter\Encryption\EncryptionException('Message authentication failed.');
```

Comment:
Should be pulled out into Encryption class and localized.

Reply:
agreed